<a href="https://colab.research.google.com/github/Aliasgar24/bangalore-tech-salary-decoder/blob/main/LiveProject_Salary_Decoder_Aliasgar_Vandeli.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Bangalore Tech Salary Decoder

### Live Project - Unlox Academy

Built by - Aliasgar Vandeli


In [40]:
# Import the required libraries

import pandas as pd
import numpy as np

In [41]:
# Load the salary dataset

df = pd.read_csv("bangalore_tech_salaries.csv")

# Display the first five rows

df.head()

,Employee_ID,Role,years_exp,Current_CTC,Previous_CTC,Company,company_TYPE,Skills,Location,Education_Tier,Joining_Year,Work_Mode
0,BLR0065,DS,6,49.4,32.4 LPA,Ola,Unicorn,"JIRA, JavaScript, Spring Boot, Figma, NumPy, A...",HSR Layout,Tier-1,2022,Remote
1,BLR0080,DevOps,0,9.7,NaN,Walmart Labs,MNC,"Tableau, Python, Figma, JavaScript, TensorFlow",Electronic City,2,2024,Remote
2,BLR0166,SDE Backend,6,"3,360,000",24.2 LPA,Postman,Mid-size,"Agile, Excel, GCP, MongoDB, Java, ML",Whitefield,Tier-1,2023,Hybrid
3,BLR0219,SDE Frontend,8,41.4 LPA,30.3,Amazon India,MNC,"PyTorch, Redis, System Design",MG Road,Tier-2,2022,Work from Office
4,BLR0491,Product Lead,5,"3,640,000",30.6 LPA,QuickPay AI,early-stage,"Tableau, Deep Learning, Agile",Whitefield,Tier-2,2023,Onsite


In [42]:
# Checking the dataset

print("Shape of Dataset:", df.shape)
print()

print("Column Names:")
print(df.columns)
print()

print("Data Types:")
print(df.dtypes)

Shape of Dataset: (1015, 12)

Column Names:
Index(['Employee_ID', 'Role ', 'years_exp', 'Current_CTC', 'Previous_CTC',
       'Company', 'company_TYPE', 'Skills', 'Location', 'Education_Tier',
       'Joining_Year', 'Work_Mode'],
      dtype='object')

Data Types:
Employee_ID       object
Role              object
years_exp          int64
Current_CTC       object
Previous_CTC      object
Company           object
company_TYPE      object
Skills            object
Location          object
Education_Tier    object
Joining_Year       int64
Work_Mode         object
dtype: object


In [43]:
# Checking missing values and duplicate records

print("Missing Values")
print(df.isnull().sum())
print()

print("Duplicate Rows:", df.duplicated().sum())

Missing Values
Employee_ID         0
Role                0
years_exp           0
Current_CTC         0
Previous_CTC      200
Company             0
company_TYPE        0
Skills             27
Location           20
Education_Tier      0
Joining_Year        0
Work_Mode           0
dtype: int64

Duplicate Rows: 15


In [44]:
# General information about the dataset

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1015 entries, 0 to 1014
Data columns (total 12 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   Employee_ID     1015 non-null   object
 1   Role            1015 non-null   object
 2   years_exp       1015 non-null   int64 
 3   Current_CTC     1015 non-null   object
 4   Previous_CTC    815 non-null    object
 5   Company         1015 non-null   object
 6   company_TYPE    1015 non-null   object
 7   Skills          988 non-null    object
 8   Location        995 non-null    object
 9   Education_Tier  1015 non-null   object
 10  Joining_Year    1015 non-null   int64 
 11  Work_Mode       1015 non-null   object
dtypes: int64(2), object(10)
memory usage: 95.3+ KB


In [45]:
# Statistical Summary

df.describe()

,years_exp,Joining_Year
count,1015.000000,1015.000000
mean,3.429557,2022.698522
std,2.558721,1.308351
min,0.000000,2020.000000
25%,1.000000,2022.000000
50%,3.000000,2023.000000
75%,5.000000,2024.000000
max,12.000000,2024.000000


In [46]:
# After this task-1
# The dataset has 1015 rows and 12 columns
# The Previous_ctc column has 200 missing values, skills column has 27 missing values and Location column has 20
# all other columns are complete
# the dataset has 15 duplicate rows
# The dataset requires cleaning

# Task 2 - Data Cleaning

In [47]:
# Rename the columns, this will make the columns in same format, like all lowercase letters

df.columns = df.columns.str.strip()
df.columns = df.columns.str.lower()

print(df.columns)

Index(['employee_id', 'role', 'years_exp', 'current_ctc', 'previous_ctc',
       'company', 'company_type', 'skills', 'location', 'education_tier',
       'joining_year', 'work_mode'],
      dtype='object')


In [48]:
# Now check for unique roles

df['role'].unique()

array(['DS', 'DevOps', 'SDE Backend', 'SDE Frontend', 'Product Lead',
       'Product Manager', 'Site Reliability Engineer', 'DA', 'UI/UX',
       'SDE-Frontend', 'UI Designer', 'ML Engineer', 'Frontend Dev',
       'Data Scientist', 'Full Stack Engineer', 'Designer',
       'Product Designer', 'BA', 'Frontend Engineer', 'BI Analyst',
       'Backend Dev', 'Sr PM', 'FE', 'SDE-Backend', 'Data Analyst',
       'Business Analyst', 'Backend Developer',
       'Business Systems Analyst', 'Data Science Engineer', 'FullStack',
       'BE', 'Frontend Developer', 'Backend Engineer', 'UX Designer',
       'Infra Engineer', 'PM', 'SDE FS', 'DevOps Engineer',
       'Analytics Engineer', 'Fullstack Dev', 'SDE Full-Stack', 'SRE'],
      dtype=object)

In [49]:
# assign diffrent roles a comman role name

role_mapping = {

    # Data Scientist
    "DS": "Data Scientist",

    # DevOps
    "DevOps": "DevOps Engineer",
    "DevOps Engineer": "DevOps Engineer",

    # Backend
    "Backend Dev": "SDE Backend",
    "Backend Developer": "SDE Backend",
    "Backend Engineer": "SDE Backend",
    "SDE-Backend": "SDE Backend",
    "BE": "SDE Backend",

    # Frontend
    "Frontend Dev": "SDE Frontend",
    "Frontend Engineer": "SDE Frontend",
    "FE": "SDE Frontend",
    "SDE-Frontend": "SDE Frontend",
    "Frontend Developer": "SDE Frontend",

    # Full Stack
    "Full Stack Engineer": "SDE Full-Stack",
    "FullStack": "SDE Full-Stack",
    "Fullstack Dev": "SDE Full-Stack",

    # Product
    "Product Lead": "Product Manager",
    "PM": "Product Manager",

    # UI/UX
    "UI Designer": "UI/UX Designer",
    "Designer": "UI/UX Designer",
    "UX Designer": "UI/UX Designer",
    "UI/UX": "UI/UX Designer",

    # Data Analyst
    "DA": "Data Analyst",

    # Business Analyst
    "BA": "Business Analyst",
    "BI Analyst": "Business Analyst",
    "Business Systems Analyst": "Business Analyst",

    # Site Reliability Engineer
    "SRE": "Site Reliability Engineer",

    # Others
    "ML Engineer": "ML Engineer",
    "Data Science Engineer": "Data Science Engineer",
    "Analytics Engineer": "Analytics Engineer",
    "Infra Engineer": "Infra Engineer",
    "SDE FS": "SDE Full-Stack",
    "Sr PM": "Senior Product Manager"
}

df["role_clean"] = df["role"].replace(role_mapping)

df["role_clean"].value_counts()

,count
role_clean,
Business Analyst,140
SDE Backend,116
SDE Frontend,105
SDE Full-Stack,97
UI/UX Designer,93
Product Manager,87
Data Scientist,66
Site Reliability Engineer,61
Data Analyst,51


In [50]:
# Lets check unique values for company type

df["company_type"].unique()

array(['Unicorn', 'MNC', 'Mid-size', 'early-stage', 'Early-stage',
       'EARLY-STAGE', 'mnc', 'unicorn', 'MID-SIZE', 'mid-size', 'UNICORN'],
      dtype=object)

In [51]:
# Lets group together the similar ones

company_mapping = {
    "Unicorn": "Unicorn",
    "unicorn": "Unicorn",
    "UNICORN": "Unicorn",

    "MNC": "MNC",
    "mnc": "MNC",

    "Mid-size": "Mid-size",
    "mid-size": "Mid-size",
    "MID-SIZE": "Mid-size",

    "Early-stage": "Early-stage",
    "early-stage": "Early-stage",
    "EARLY-STAGE": "Early-stage"
}

df["company_type_clean"] = df["company_type"].replace(company_mapping)

df["company_type_clean"].value_counts()

,count
company_type_clean,
MNC,348
Mid-size,261
Unicorn,207
Early-stage,199


In [52]:
# Now for education Tier

df["education_tier"].unique()

array(['Tier-1', '2', 'Tier-2', 'Tier 1', 'Tier 2', 'T3', 'Tier-3', 'T2',
       '1', 'T1', 'Tier 3', '3'], dtype=object)

In [53]:
# Lets group them together

education_mapping = {
    "Tier-1": "Tier 1",
    "Tier 1": "Tier 1",
    "T1": "Tier 1",
    "1": "Tier 1",

    "Tier-2": "Tier 2",
    "Tier 2": "Tier 2",
    "T2": "Tier 2",
    "2": "Tier 2",

    "Tier-3": "Tier 3",
    "Tier 3": "Tier 3",
    "T3": "Tier 3",
    "3": "Tier 3"
}

df["education_tier_clean"] = df["education_tier"].replace(education_mapping)

df["education_tier_clean"].value_counts()

,count
education_tier_clean,
Tier 2,510
Tier 3,309
Tier 1,196


In [54]:
# Now we check the diff formats used in current_ctc

df["current_ctc"].head(20)

,current_ctc
0,49.4
1,9.7
2,"3,360,000"
3,41.4 LPA
4,"3,640,000"
5,32.1
6,24.6
7,21.7 LPA
8,21.0
9,26.9 LPA


In [55]:
# Now we need to clean the current_ctc and bring it into a single float format (LPA)



def parse_ctc(value):

    if pd.isna(value):
        return np.nan

    value = str(value)

    # Now we Remove unnecessary symbols like rupee, LPA, the commas
    value = value.replace("₹", "")
    value = value.replace("LPA", "")
    value = value.replace(",", "")
    value = value.strip()

    salary = float(value)

    # Convert rupees into LPA
    if salary > 100000:
        salary = salary / 100000

    return salary


df["current_ctc_lpa"] = df["current_ctc"].apply(parse_ctc)

print(df["current_ctc_lpa"].head(20))

0     49.4
1      9.7
2     33.6
3     41.4
4     36.4
5     32.1
6     24.6
7     21.7
8     21.0
9     26.9
10    28.7
11    13.0
12    28.0
13    10.0
14    14.2
15    15.5
16    17.0
17    17.5
18    10.4
19    17.1
Name: current_ctc_lpa, dtype: float64


In [56]:
# As we see, all salaries are converted to a single format

In [57]:
# Now we Convert Previous_CTC into float (LPA), we use the functoni created above def parse_ctc and pass the previous_ctc values

df["previous_ctc_lpa"] = df["previous_ctc"].apply(parse_ctc)

print(df["previous_ctc_lpa"].head(20))

0     32.4
1      NaN
2     24.2
3     30.3
4     30.6
5     22.0
6     19.3
7     17.7
8      NaN
9     18.9
10    17.4
11     NaN
12    18.5
13     NaN
14     8.7
15    11.8
16    12.2
17    10.7
18     NaN
19    11.3
Name: previous_ctc_lpa, dtype: float64


In [58]:
# Fill missing values in Skills and Location

df["skills"] = df["skills"].fillna("Unknown")
df["location"] = df["location"].fillna("Unknown")

In [59]:
# Remove duplicate rows from the dataset

df = df.drop_duplicates()

print("Number of rows after removing duplicates:", len(df))

Number of rows after removing duplicates: 1000


In [60]:
# the dataset previuosly had 1015 rows with 15 duplicates, which are now removed

In [61]:
# Now lets Verify the cleaned dataset

print("Data Types")
print(df.dtypes)

print("\nRole Counts")
print(df["role_clean"].value_counts())

print("\nCompany Type Counts")
print(df["company_type_clean"].value_counts())

print("\nEducation Tier Counts")
print(df["education_tier_clean"].value_counts())

Data Types
employee_id              object
role                     object
years_exp                 int64
current_ctc              object
previous_ctc             object
company                  object
company_type             object
skills                   object
location                 object
education_tier           object
joining_year              int64
work_mode                object
role_clean               object
company_type_clean       object
education_tier_clean     object
current_ctc_lpa         float64
previous_ctc_lpa        float64
dtype: object

Role Counts
role_clean
Business Analyst             136
SDE Backend                  114
SDE Frontend                 105
SDE Full-Stack                96
UI/UX Designer                90
Product Manager               86
Data Scientist                64
Site Reliability Engineer     61
Data Analyst                  50
DevOps Engineer               46
ML Engineer                   31
Data Science Engineer         30
Senior Prod

In [62]:
# Now lets Check missing values after cleaning

print(df.isnull().sum())

employee_id               0
role                      0
years_exp                 0
current_ctc               0
previous_ctc            194
company                   0
company_type              0
skills                    0
location                  0
education_tier            0
joining_year              0
work_mode                 0
role_clean                0
company_type_clean        0
education_tier_clean      0
current_ctc_lpa           0
previous_ctc_lpa        194
dtype: int64


In [63]:
# We did not remove missing values of previous_ctc as they can be of frehsers having no salary

# Task 3.1 - CTC Distribution by Role

In [64]:
# Calculate salary statistics for each role

role_salary = df.groupby("role_clean")["current_ctc_lpa"].agg(
    ["median", "mean", "min", "max"]
)

# Sort roles by median salary in descending order

role_salary = role_salary.sort_values(by="median", ascending=False)

print(role_salary)

                           median       mean   min   max
role_clean                                              
Senior Product Manager      31.50  35.400000  10.8  65.5
Product Manager             31.30  33.747674  12.0  80.1
ML Engineer                 27.50  30.335484  10.0  64.5
Data Science Engineer       25.65  28.356667  14.3  75.6
Infra Engineer              23.45  22.610000   9.2  45.5
Site Reliability Engineer   23.30  23.767213   8.8  55.0
Data Scientist              22.85  25.879687  10.9  73.5
SDE Full-Stack              22.35  25.409375   8.9  71.7
Analytics Engineer          21.40  20.909091   6.8  37.9
SDE Frontend                21.20  22.193333   6.7  84.4
SDE Backend                 21.10  23.284210   7.9  55.1
DevOps Engineer             19.45  22.891304   9.7  60.3
Product Designer            19.00  21.526087  10.9  55.7
UI/UX Designer              18.70  20.825555   6.2  63.3
Business Analyst            18.30  20.825000   5.2  52.7
Data Analyst                16.

In [65]:
# highest and lowest paying roles based on median salary

highest_role = role_salary.index[0]
lowest_role = role_salary.index[-1]

print("Highest Paying Role:", highest_role)
print("Lowest Paying Role:", lowest_role)

Highest Paying Role: Senior Product Manager
Lowest Paying Role: Data Analyst


In [66]:
# Compare mean and median salary for each role

print("Mean vs Median Salary by Role")
print(role_salary[["median", "mean"]])

Mean vs Median Salary by Role
                           median       mean
role_clean                                  
Senior Product Manager      31.50  35.400000
Product Manager             31.30  33.747674
ML Engineer                 27.50  30.335484
Data Science Engineer       25.65  28.356667
Infra Engineer              23.45  22.610000
Site Reliability Engineer   23.30  23.767213
Data Scientist              22.85  25.879687
SDE Full-Stack              22.35  25.409375
Analytics Engineer          21.40  20.909091
SDE Frontend                21.20  22.193333
SDE Backend                 21.10  23.284210
DevOps Engineer             19.45  22.891304
Product Designer            19.00  21.526087
UI/UX Designer              18.70  20.825555
Business Analyst            18.30  20.825000
Data Analyst                16.30  17.150000




Senior Product Managers have the highest median salary (31.5 LPA), while Data Analysts have the lowest (16.3 LPA).

# Task 3.2 - Experience Curve for SDE Backend

In [70]:
# Select only SDE Backend employees

backend_df = df[df["role_clean"] == "SDE Backend"].copy()

print("Total SDE Backend Employees:", len(backend_df))

Total SDE Backend Employees: 114


In [71]:
# Divide employees into experience groups

bins = [-1, 1, 3, 5, 100]
labels = ["0-1", "2-3", "4-5", "6+"]

backend_df["exp_band"] = pd.cut(
    backend_df["years_exp"],
    bins=bins,
    labels=labels
)

print(backend_df[["years_exp", "exp_band"]].head())

    years_exp exp_band
2           6       6+
23          1      0-1
34          4      4-5
42          3      2-3
44          3      2-3


In [72]:
# Calculate median salary for each experience band

experience_salary = backend_df.groupby("exp_band")["current_ctc_lpa"].median()

print(experience_salary)

exp_band
0-1    11.65
2-3    20.00
4-5    25.85
6+     40.40
Name: current_ctc_lpa, dtype: float64


/tmp/ipykernel_1953/613596423.py:3: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  experience_salary = backend_df.groupby("exp_band")["current_ctc_lpa"].median()


In [73]:
# Calculate salary growth for each experience band

growth_1 = ((experience_salary["2-3"] - experience_salary["0-1"]) / experience_salary["0-1"]) * 100

growth_2 = ((experience_salary["4-5"] - experience_salary["2-3"]) / experience_salary["2-3"]) * 100

growth_3 = ((experience_salary["6+"] - experience_salary["4-5"]) / experience_salary["4-5"]) * 100

print("Growth from 0-1 to 2-3 years :", round(growth_1, 2), "%")
print("Growth from 2-3 to 4-5 years :", round(growth_2, 2), "%")
print("Growth from 4-5 to 6+ years  :", round(growth_3, 2), "%")

Growth from 0-1 to 2-3 years : 71.67 %
Growth from 2-3 to 4-5 years : 29.25 %
Growth from 4-5 to 6+ years  : 56.29 %




The median salary of SDE Backend professionals increases steadily with experience, from 11.65 LPA for 0–1 years to 40.40 LPA for professionals with more than 6 years of experience.

# Task 3.3 - Skill Premium for SDEs

In [74]:
# Select all SDE roles

sde_df = df[
    (df["role_clean"] == "SDE Backend") |
    (df["role_clean"] == "SDE Frontend") |
    (df["role_clean"] == "SDE Full-Stack")
].copy()

print("Total SDE Employees:", len(sde_df))

Total SDE Employees: 315


In [75]:
# List of skills to analyse, we will select these 4 skills

skills = ["AWS", "ML", "System Design", "Kubernetes"]

In [76]:
# Compare median salary with and without each skill

for skill in skills:

    with_skill = sde_df[
        sde_df["skills"].str.contains(skill, case=False, na=False)
    ]["current_ctc_lpa"].median()

    without_skill = sde_df[
        ~sde_df["skills"].str.contains(skill, case=False, na=False)
    ]["current_ctc_lpa"].median()

    premium = ((with_skill - without_skill) / without_skill) * 100

    print("Skill :", skill)
    print("Median with skill :", round(with_skill, 2), "LPA")
    print("Median without skill :", round(without_skill, 2), "LPA")
    print("Premium :", round(premium, 2), "%")
    print("-" * 40)

Skill : AWS
Median with skill : 20.95 LPA
Median without skill : 21.9 LPA
Premium : -4.34 %
----------------------------------------
Skill : ML
Median with skill : 23.95 LPA
Median without skill : 21.3 LPA
Premium : 12.44 %
----------------------------------------
Skill : System Design
Median with skill : 25.3 LPA
Median without skill : 21.0 LPA
Premium : 20.48 %
----------------------------------------
Skill : Kubernetes
Median with skill : 23.2 LPA
Median without skill : 21.5 LPA
Premium : 7.91 %
----------------------------------------


among the skills we chose, System Design provides highest salary premium (20.48%), whereas AWS provides a negative premium of (-4.34%)

# Task 3.4 - Company Type Premium

In [77]:
# Select only SDE Backend employees

backend_df = df[df["role_clean"] == "SDE Backend"].copy()

In [78]:
# Calculate median salary for each company type

company_salary = backend_df.groupby("company_type_clean")["current_ctc_lpa"].median()

print(company_salary)

company_type_clean
Early-stage    18.60
MNC            20.45
Mid-size       19.50
Unicorn        27.35
Name: current_ctc_lpa, dtype: float64


In [80]:
# Calculate Unicorn premium over MNC

unicorn_salary = company_salary["Unicorn"]
mnc_salary = company_salary["MNC"]

premium = ((unicorn_salary - mnc_salary) / mnc_salary) * 100

print("Unicorn Median Salary :", round(unicorn_salary, 2))
print("MNC Median Salary :", round(mnc_salary, 2))
print("Premium :", round(premium, 2), "%")

Unicorn Median Salary : 27.35
MNC Median Salary : 20.45
Premium : 33.74 %


For SDE Backend Employees, Unicorn provides a median salary of (27.35), MNC on other hand provides (20.45)

# Task 3.5 - Underpaid Professionals

In [81]:
# Create experience bands for the entire dataset

bins = [-1, 1, 3, 5, 100]
labels = ["0-1", "2-3", "4-5", "6+"]

df["exp_band"] = pd.cut(
    df["years_exp"],
    bins=bins,
    labels=labels
)

print(df[["years_exp", "exp_band"]].head())

   years_exp exp_band
0          6       6+
1          0      0-1
2          6       6+
3          8       6+
4          5      4-5


In [82]:
# Keep  groups that have at least 10 employees

valid_groups = df.groupby(
    ["role_clean", "company_type_clean", "exp_band"]
).filter(lambda x: len(x) >= 10)

print("Valid Rows:", len(valid_groups))

Valid Rows: 184


/tmp/ipykernel_1953/4187418558.py:3: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  valid_groups = df.groupby(


In [83]:
# Calculate the median salary of each group

valid_groups["group_median"] = valid_groups.groupby(
    ["role_clean", "company_type_clean", "exp_band"]
)["current_ctc_lpa"].transform("median")

/tmp/ipykernel_1953/1964420065.py:3: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  valid_groups["group_median"] = valid_groups.groupby(


In [84]:
# Calculate salary gap from the group median

valid_groups["gap"] = (
    valid_groups["current_ctc_lpa"] -
    valid_groups["group_median"]
)

In [85]:
# Display the top 10 most underpaid professionals

underpaid = valid_groups.sort_values(by="gap").head(10)

print(
    underpaid[
        [
            "employee_id",
            "role_clean",
            "company_type_clean",
            "years_exp",
            "current_ctc_lpa",
            "group_median",
            "gap"
        ]
    ]
)

    employee_id        role_clean company_type_clean  years_exp  \
349     BLR0954  Business Analyst                MNC          6   
904     BLR0925       SDE Backend                MNC          4   
250     BLR0072       SDE Backend                MNC          4   
728     BLR0301   Product Manager                MNC          2   
249     BLR0733  Business Analyst                MNC          7   
753     BLR0611   Product Manager                MNC          2   
688     BLR0203       SDE Backend                MNC          4   
711     BLR0451      SDE Frontend                MNC          4   
858     BLR0033      SDE Frontend                MNC          4   
538     BLR0486   Product Manager                MNC          2   

     current_ctc_lpa  group_median    gap  
349             25.4         37.65 -12.25  
904             19.4         27.30  -7.90  
250             19.5         27.30  -7.80  
728             24.7         31.45  -6.75  
249             31.1         37.65  -6.55 



The most underpaid employee in the filtered groups is BLR0954 (Business Analyst), earning 25.40 LPA, which is 12.25 LPA below the median salary of their peer group.

# Task 4 - Final Printed Report

In [87]:
# Print the final salary report

print("=" * 60)
print("           BANGALORE TECH SALARY DECODER")
print("           Built by Aliasgar Vandeliwala")
print("=" * 60)

print("\nDataset : 1000 Bengaluru Tech Professionals")
print("Period  : 2024 Employment Snapshot")

print("\n" + "-" * 60)
print("MEDIAN CTC BY ROLE")
print("-" * 60)

for role, row in role_salary.iterrows():
    print(f"{role:<30} {row['median']:.2f} LPA")

print("\n" + "-" * 60)
print("SDE BACKEND CTC BY EXPERIENCE")
print("-" * 60)

print(f"0-1 Years : {experience_salary['0-1']:.2f} LPA")
print(f"2-3 Years : {experience_salary['2-3']:.2f} LPA (+{growth_1:.2f}%)")
print(f"4-5 Years : {experience_salary['4-5']:.2f} LPA (+{growth_2:.2f}%)")
print(f"6+ Years  : {experience_salary['6+']:.2f} LPA (+{growth_3:.2f}%)")

print("\n" + "-" * 60)
print("COMPANY TYPE PREMIUM (SDE Backend)")
print("-" * 60)

for company, salary in company_salary.items():
    print(f"{company:<15} {salary:.2f} LPA")

print(f"\nUnicorn Premium over MNC : {premium:.2f}%")

print("\n" + "-" * 60)
print("TOP 5 MOST UNDERPAID PROFESSIONALS")
print("-" * 60)

top5 = underpaid.head(5)

for index, row in top5.iterrows():
    print(
        f"{row['employee_id']}  "
        f"{row['role_clean']:<20} "
        f"{row['company_type_clean']:<10} "
        f"{row['years_exp']} yrs   "
        f"Gap : {row['gap']:.2f} LPA"
    )

print("=" * 60)

           BANGALORE TECH SALARY DECODER
           Built by Aliasgar Vandeliwala

Dataset : 1000 Bengaluru Tech Professionals
Period  : 2024 Employment Snapshot

------------------------------------------------------------
MEDIAN CTC BY ROLE
------------------------------------------------------------
Senior Product Manager         31.50 LPA
Product Manager                31.30 LPA
ML Engineer                    27.50 LPA
Data Science Engineer          25.65 LPA
Infra Engineer                 23.45 LPA
Site Reliability Engineer      23.30 LPA
Data Scientist                 22.85 LPA
SDE Full-Stack                 22.35 LPA
Analytics Engineer             21.40 LPA
SDE Frontend                   21.20 LPA
SDE Backend                    21.10 LPA
DevOps Engineer                19.45 LPA
Product Designer               19.00 LPA
UI/UX Designer                 18.70 LPA
Business Analyst               18.30 LPA
Data Analyst                   16.30 LPA

---------------------------------------

# Task 5 - Key Insights

 Insight 1

Senior Product Managers have the highest median salary of (31.50 LPA), while Data Analysts have the lowest median salary of (16.30 LPA) . This shows that managerial roles generally offer significantly higher compensation than entry-level  roles.

Insight 2

The median salary of SDE Backend professionals increases from (11.65 LPA) for employees with 0–1 years of experience to 40.40 LPA for those with 6+ years of experience. This implies the impact of experience on salary growth in software engineering

Insight 3

Among the selected technical skills, System Design provides the highest salary premium of 20.48%. For SDE Backend professionals, Unicorn companies also offer a 33.74% higher median salary than MNCs.